### ISSUE TITLE
Inconsistent behavior in tf.contrib.distributions.percentile for NaN values

### ISSUE THREAD

**Issue by javidcf** [tensorflow/tensorflow#22481]
### System information
- **Have I written custom code (as opposed to using a stock example script provided in TensorFlow)**: Yes
- **OS Platform and Distribution (e.g., Linux Ubuntu 16.04)**: Windows 10
- **Mobile device (e.g. iPhone 8, Pixel 2, Samsung Galaxy) if the issue happens on mobile device**: NA
- **TensorFlow installed from (source or binary)**: Binary
- **TensorFlow version (use command below)**: v1.10.0-rc1-19-g656e7a2b34 1.10.0
- **Python version**: 3.6.6 (Anaconda)
- **Bazel version (if compiling from source)**: NA
- **GCC/Compiler version (if compiling from source)**: NA
- **CUDA/cuDNN version**: NA
- **GPU model and memory**: NA
- **Exact command to reproduce**: NA

### Describe the problem
[`tf.contrib.distributions.percentile`](https://www.tensorflow.org/api_docs/python/tf/contrib/distributions/percentile) behaves inconsistently when NaN values are given. While the documentation does not explicitly states how this case should be handled, one would normally expect to always receive NaN, or maybe ignore them in the computation (compute the percentile of non-NaN values), or even replace them with 0. But the current logic seems to somehow depend on the order of the values in the input: trying to compute the median of `[1.0, NaN]` results in `1.0`, while for `[NaN, 1.0]` it is `NaN`.

The behavior seems to be different also between CPU and GPU. I obtained the results above while running on CPU. The same test on GPU (CUDA 9.0, CUDNN 7, Titan V) produced `NaN` in both cases; however, the computed median on GPU for `[1.0, 1.0, NaN]` was again `1.0`.

### Source code / logs

```py
import tensorflow as tf
import math

x = tf.placeholder(tf.float32, [None])
m = tf.contrib.distributions.percentile(x, q=50., validate_args=True)
with tf.Session() as sess:
    print(sess.run(m, feed_dict={ x: [1, math.nan] }))  # Prints 1.0
    print(sess.run(m, feed_dict={ x: [math.nan, 1] }))  # Prints nan
```

**Comment by axch** [comment]
`tf.contrib.distributions` are deprecated in TensorFlow; development has migrated to https://github.com/tensorflow/probability/tree/master/tensorflow_probability/python/distributions, including `sample_stats.py`.  This issue will be addressed there.
 
I presume the source of the problem is that percentile relies on sort+index, but NaN values do not sort consistently: both NaN < x and x < NaN are always false, for all x (including x = NaN).  One candidate resolution would be to explicitly check the input for NaN, and return NaN in that case.  @jvdillon, opinions on setting the interface of `percentile` to that?

**Comment by jvdillon** [comment]
Always tough to decide how to handle NaN.  I think the best thing would be to add `validate_args` which when true would enable an assertion which fires on detection of NaN.

I think we should be hesitant to more aggressive pruning since this would potentially incur a cost whether or not you actually had NaNs. (And TFP has a strict "only pay for what you eat" policy.)

Thoughts?

**Comment by axch** [comment]
I don't think the cost is that bad: The function already does a complete sort, so the NaN check should be asymptotically cheaper.  There shouldn't be any control flow, either: compute a mask of which batch members have NaNs, then tf.where on that mask to select either NaN or the sort+index result.

I don't think this is a job for `validate_args` -- having a NaN argument isn't invalid, it just means a NaN answer.  Maybe instead add `allow_nan_stats=True` and raise an exception if there is a NaN in the input and this flag is false?

**Comment by jvdillon** [comment]
`allow_nan_stats=True` sgtm.  However our policy is pretty strict, asymptotic costs or otherwise. We could add another percentile function which masks `NaN`, however that seems weirdly specific. Can you make a case why this should be handled on the TFP side? Ie, a user can do this today:

   percentile(tfp.where(tf.isnan(x),tf.zeros_like(x),x))

The advantage of this is its fully self-documenting and nearly as terse as if percentile took a mask arg.

**Comment by axch** [comment]
Just to make sure we are both on the same page, I am not talking about masking input NaNs (which, I agree, would be weirdly specific) but about propagating them.  The interface I think we should implement is the same as numpy:
```
>>> np.percentile([float("nan"), 1.0], 50)
nan
>>> np.percentile([1.0, float("nan")], 50)
nan
>>> np.percentile([1.0, 2.0, 3.0, 4.0, float("nan")], 20)
nan
>>> np.percentile([1.0, 2.0, 3.0, 4.0], 20)
1.6000000000000001
>>> np.percentile([1.0, float("nan"), 2.0, 3.0, 4.0], 20)
nan
```

This can, of course, be done in user-land:
```
def nan_propagating_percentile(x, q, axis):
  nan_batch_members = tf.reduce_any(tf.isnan(x), axis=axis)
  return tf.where(
    nan_batch_members,
    np.nan * tf.ones(tf.shape(nan_batch_members), dtype=x.dtype),
    percentile(x, q, axis=axis))
```

but that seems like something of an imposition.  @jvdillon: How do you value the cost of `is_nan` + `reduce_any` + `where` (for an operation that already includes a sort and slice) in comparison with the benefit of API compatibility, and of avoiding giving non-`NaN` answers on `NaN` inputs?

**Comment by tensorflowbutler** [comment]
Nagging Assignee @axch: It has been 14 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by tensorflowbutler** [comment]
Nagging Assignee @axch: It has been 29 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by tensorflowbutler** [comment]
Nagging Assignee @axch: It has been 44 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by tensorflowbutler** [comment]
Nagging Assignee @axch: It has been 59 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.
